### PR TITLE
formal: add btormc engine as a backend option

### PR DESCRIPTION
**Pull request by ekiwi**
This PR adds support for the `btormc` model checker which can be substantially faster than using Z3 or CVC4.

It also improves the CI for our tests of the formal verification support: Now (almost) every test is executed with each one of the supported backends. This CI improvement revealed a bug in how we parse responses from the CVC4 solver so I had to add the fix to this PR.

### CODE REVIEW

File: src/test/scala/chiseltest/formal/FormalBackendOption.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    configMap.get("formal_engine") match {
+      case Some("z3") => Z3EngineAnnotation
+      case Some("cvc4") => CVC4EngineAnnotation
+      case Some("btormc") => BtormcEngineAnnotation
+      case Some(unknown) => throw new RuntimeException(s"Unknown formal engine: $unknown")
+      // default
+      case None => Z3EngineAnnotation
+    }
+  }
+}
```
Comment (chick) [inline src/test/scala/chiseltest/formal/FormalBackendOption.scala:23]: Missing newline at end

**Review by chick — APPROVED** [review]
I left a couple nit comment I put in there, but otherwise this looks fine to me.

File: src/main/scala/chiseltest/formal/backends/btor/Btor2WitnessParser.scala [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      val (value, index) = if (is_array) {
+        assert(parts(1).startsWith("[") && parts(1).endsWith("]"), s"Expected `[index]`, not `${parts(1)}` in: $parts")
+        val indexString = parts(1).drop(1).dropRight(1)
+        val ii = if (indexString == "*") None else Some(BigInt(indexString, 2))
+        (BigInt(parts(2), 2), ii)
+      } else { (BigInt(parts(1), 2), None) }
+      Assignment(ii, value, index, symbol = parts.last, is_array)
+    }
+
+    def parse_line(line: String): Unit = {
```
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/chiseltest/formal/backends/btor/Btor2WitnessParser.scala]: thanks

File: src/main/scala/chiseltest/formal/backends/btor/Btor2WitnessParser.scala [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+          }
+        }
+      }
+      //println(s" -> $state")
+    }
+
+    breakable {
+      lines.foreach { ll =>
+        //println(ll.trim)
+        parse_line(ll.trim)
```
[the remark this answers is not in the document]
Comment (ekiwi) [reply src/main/scala/chiseltest/formal/backends/btor/Btor2WitnessParser.scala]: fixed

File: src/test/scala/chiseltest/formal/FormalBackendOption.scala (hunk 1 above)
Comment (ekiwi) [reply src/test/scala/chiseltest/formal/FormalBackendOption.scala]: fixed

**Pull request merged**